src/orion/core/worker/warm_start/knowledge_base.py

```diff
@@ -0,0 +1,79 @@
+""" Interface for the Knowledge Base, which is not currently in the Orion codebase.
+"""
+from __future__ import annotations
+
+import typing
+from abc import ABC, abstractmethod
+from logging import getLogger
+from typing import Container
+
+
+if typing.TYPE_CHECKING:
+    from orion.client import ExperimentClient
+    from orion.core.worker.experiment import Experiment
+    from orion.core.worker.trial import Trial
+    from orion.storage.base import BaseStorageProtocol
+
+
+log = getLogger(__file__)
+
+from .experiment_config import ExperimentInfo
+
+
+class KnowledgeBase(ABC, Container[ExperimentInfo]):
+    """Abstract Base Class for the KnowledgeBase, which currently isn't part of the
+    Orion codebase.
+    """
+
+    @abstractmethod
+    def get_related_trials(
+        self,
+        target_experiment: Experiment | ExperimentClient,
+        max_trials: int | None = None,
+    ) -> list[tuple[ExperimentInfo, list[Trial]]]:
+        """Retrieve experiments 'similar' to `target_experiment` and their trials.
+
+        When `max_trials` is given, only up to `max_trials` are returned in total for
+        all experiments.
+
+        Parameters
+        ----------
+        target_experiment : Union[Experiment, ExperimentClient]
+            The target experiment, or experiment client.
+        max_trials : int, optional
+            Maximum total number of trials to fetch. By default `None`, in which case
+            all trials from all 'related' experiments are returned.
+
+        Returns
+        -------
+        list[tuple[ExperimentInfo, list[Trial]]]
+            A list of tuples of experiments and the similar trials from that experiment.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def n_stored_experiments(self) -> int:
+        """Returns the current number of experiments registered the Knowledge base."""
+
+    @abstractmethod
+    def add_experiment(self, experiment: Experiment | ExperimentClient) -> None:
```

Comment:
Interesting! What do you foresee as being hard? Would you mind giving me an example, just so I can understand better?

Wouldn't it be harder to remember which experiments are in which storage objects?

E.g.
```python
knowledge_base = KnowledgeBase()
my_previous_experiment = (...)
knowledge_base.add_experiment(my_previous_experiment)
target_experiment = create_experiment("target", ..., knowledge_base=knowledge_base)
```

Seems more explicit to me than this:

```python
knowledge_base = KnowledgeBase()
knowledge_base.add_storage(PickleDb("my_previous_experiment_storage.pkl"))
target_experiment = create_experiment("target", ..., knowledge_base=knowledge_base)
```
Since it's not necessarily the case that the storage has only one experiment.

Reply:
It is more explicit indeed but then I can only reproduce the experiment using the same code (or re-implementation) to add experiments piece by piece, whereas it could be reproduced based on a configuration of storage configs. But even with the storage it can be problematic since they are mutable... There is something I don't like here for reproducibility :sweat_smile: Maybe the KB should always lead to a versionned dataset based on current status of the storages, and then we can reproduce by used the versionned dataset instead of the original storage configs.